[BPF] Generate BTF info using 'btf:type_tag' annotation

eddyz87 opened the pull request:
This is a follow-up for BPF mailing list discussion at [1].

Previous commit in a series updated DWARF generation for the following example:

    int __attribute__((btf_type_tag("tag1"))) *g;

To generate DWARF that looks as follows:

    0x0000001e:   DW_TAG_variable
                    DW_AT_name      ("g")
                    DW_AT_type      (0x00000029 "int *")

    0x00000029:   DW_TAG_pointer_type
                    DW_AT_type      (0x00000032 "int")

    0x00000032:   DW_TAG_base_type
                    DW_AT_name      ("int")

    0x00000036:     DW_TAG_LLVM_annotation
                      DW_AT_name    ("btf:type_tag")
                      DW_AT_const_value     ("tag1")

The fresh part is attachment of `btf:type_tag` annotations to types other than pointers. This commit changes BTF generation to rely on `btf:type_tag` annotations to generate TYPE_TAG entries. This necessitates the following changes:
- The logic for `BTFTypeTypeTag` chains creation is moved to `BTFDebug::addType()`;
- Special logic is added to avoid duplicate BTF entries for tagged and un-tagged type variants, e.g. in the following case:

```c
    #define __tag1 __attribute__((btf_type_tag("tag1")))
    #define __tag2 __attribute__((btf_type_tag("tag2")))
  
    struct foo {};
  
    struct bar {
      struct foo __tag1 aa;
      struct foo __tag2 bb;
      struct foo cc;
    };
```

  Debug information generated for this example contains three instances of `DICompositeType(name: "foo")` with different `annotations` fields, however single BTF definition for structure "foo" should be generated. Field `BTFDebug::DIDedupMap` and method `BTFDebug::lookupType()` are responsible for this logic;
- Care is taken to avoid references to type tags in relocation entries.

[1] https://lore.kernel.org/bpf/87r0w9jjoq.fsf@oracle.com/

This was previously tracked as differential revision:
https://reviews.llvm.org/D145891

eddyz87 committed:
[DebugInfo][BPF] Add 'annotations' field for DIBasicType & DISubroutineType

Extend DIBasicType and DISubroutineType with additional field
'annotations', e.g. as below:

  !5 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed, annotations: !6)
  !6 = !{!7}
  !7 = !{!"btf:type_tag", !"tag1"}

The field would be used by BPF backend to generate DWARF attributes
corresponding to "btf_type_tag" type attributes, e.g.:

  0x00000029:   DW_TAG_base_type
                  DW_AT_name	("int")
                  DW_AT_encoding	(DW_ATE_signed)
                  DW_AT_byte_size	(0x04)

  0x0000002d:     DW_TAG_LLVM_annotation
                    DW_AT_name	("btf:type_tag")
                    DW_AT_const_value	("tag1")

Such DWARF entries would be used to generate BTF definitions by tools
like pahole [1].

Note: similar fields with similar purposes are already present in
DIDerivedType and DICompositeType.

Currently "btf_type_tag" attributes are represented in debug
information as 'annotations' fields in DIDerivedType with
DW_TAG_pointer_type tag. The annotation on a pointer corresponds to
pointee having the attributes in the final BTF.

The discussion at [2] came to conclusion, that such annotations should
apply to the annotated type itself. Hence the necessity to extend
DIBasicType & DISubroutineType types with 'annotations' field to
represent cases like below:

  int __attribute__((btf_type_tag("foo"))) bar;

[1] https://github.com/acmel/dwarves
[2] https://lore.kernel.org/bpf/87r0w9jjoq.fsf@oracle.com/

This was previously tracked as differential revision:
https://reviews.llvm.org/D143966

eddyz87 committed:
[DebugInfo][BPF] Add 'btf:type_tag' annotation in DWARF

This commit is a follow-up for BPF mailing list discussion at [1].
It changes the way `__attribute__((btf_type_tag("...")))`s are
represented in DWARF.

Prior to this commit type tags could only be attached to pointers.
Such attachments associated the tags with a pointee type.
E.g. for the following C code:

    int __attribute__((btf_type_tag("tag1"))) *g;

Generated DWARF looked as follows:

    0x0000001e:   DW_TAG_variable
                    DW_AT_name      ("g")
                    DW_AT_type      (0x00000029 "int *")

    0x00000029:   DW_TAG_pointer_type
                    DW_AT_type      (0x00000032 "int")

    0x0000002e:     DW_TAG_LLVM_annotation
                      DW_AT_name    ("btf_type_tag")
                      DW_AT_const_value     ("tag1")

    0x00000032:   DW_TAG_base_type
                    DW_AT_name      ("int")

The goal of this commit is to allow attachment of type tags to the
tagged types instead. E.g. for the same example DWARF should look as
follows:

    0x0000001e:   DW_TAG_variable
                    DW_AT_name      ("g")
                    DW_AT_type      (0x00000029 "int *")

    0x00000029:   DW_TAG_pointer_type
                    DW_AT_type      (0x00000032 "int")

    0x00000032:   DW_TAG_base_type
                    DW_AT_name      ("int")

    0x00000036:     DW_TAG_LLVM_annotation
                      DW_AT_name    ("btf:type_tag")
                      DW_AT_const_value     ("tag1")

A new tag name, `btf:type_tag`, is used so that DWARF consumers
could distinguish between old and new attachment semantics.

This feature is mostly used by Linux Kernel in combination with tool
named pahole [2]. Reasonably recent versions of pahole generate
errors (1.23, 1.24) or warnings (1.25) when `DW_TAG_LLVM_annotation`
is attached to `DW_TAG_base_type` or `DW_TAG_unspecified_type`.
Hence the `btf:type_tag` generation is controlled by a hidden option
`-mllvm -btf-type-tag-v2`. The goal is to provide a way for tooling to
work on adding support `btf:type_tag` and eventually replace
`btf_type_tag` by `btf:type_tag`, removing the above option.

The commit includes the following changes:
- Changes in debug info generation:
  - New method `DIBuilder::createAnnotationsPlaceholder()` is added,
    it creates a temporary `DIDerivedType` that plays as annotations
    placeholder while debug info metadata is being constructed;
  - New overload for `CGDebugInfo::CreateType` method is added:

      llvm::DIType *CGDebugInfo::CreateType(const BTFTagAttributedType *Ty,
                                            llvm::DIFile *Unit);

    This overload collects BTF type tags in `Ty`, creates annotations
    placeholder pointing to the base type of `Ty`, registers the
    placeholder in the `CGDebugInfo::AnnotationsPlaceholder` vector.
  - `CGDebugInfo::finalize()` is updated to do the following for each
    annotation placeholder:
    - clone underlying base type;
    - attach annotations the clone using `replaceAnnotations()` call;
    - replace all placeholder usages by a clone.
  Such scheme allows to deal with type cycles.

- Changes in AST construction:
  - `ASTContext::getBTFTagAttributedType()` is updated to ensure that
    `BTFTagAttributedType` always wraps `QualType` w/o local
    constant/volatile/restricted qualifiers. This simplifies debug info
    generation.

[1] https://lore.kernel.org/bpf/87r0w9jjoq.fsf@oracle.com/
[2] https://git.kernel.org/pub/scm/devel/pahole/pahole.git/

This was previously tracked as differential revision:
https://reviews.llvm.org/D143967

eddyz87 committed:
[BPF] Generate BTF info using 'btf:type_tag' annotation

This is a follow-up for BPF mailing list discussion at [1].

Previous commit in a series updated DWARF generation for the following
example:

    int __attribute__((btf_type_tag("tag1"))) *g;

To generate DWARF that looks as follows:

    0x0000001e:   DW_TAG_variable
                    DW_AT_name      ("g")
                    DW_AT_type      (0x00000029 "int *")

    0x00000029:   DW_TAG_pointer_type
                    DW_AT_type      (0x00000032 "int")

    0x00000032:   DW_TAG_base_type
                    DW_AT_name      ("int")

    0x00000036:     DW_TAG_LLVM_annotation
                      DW_AT_name    ("btf:type_tag")
                      DW_AT_const_value     ("tag1")

The fresh part is attachment of `btf:type_tag` annotations to types
other than pointers. This commit changes BTF generation to rely on
`btf:type_tag` annotations to generate TYPE_TAG entries.
This necessitates the following changes:
- The logic for `BTFTypeTypeTag` chains creation is moved to
  `BTFDebug::addType()`;
- Special logic is added to avoid duplicate BTF entries for tagged
  and un-tagged type variants, e.g. in the following case:

  #define __tag1 __attribute__((btf_type_tag("tag1")))
  #define __tag2 __attribute__((btf_type_tag("tag2")))

  struct foo {};

  struct bar {
    struct foo __tag1 aa;
    struct foo __tag2 bb;
    struct foo cc;
  };

  Debug information generated for this example contains three
  instances of `DICompositeType(name: "foo")` with different
  `annotations` fields, however single BTF definition for structure
  "foo" should be generated. Field `BTFDebug::DIDedupMap` and method
  `BTFDebug::lookupType()` are responsible for this logic;
- Care is taken to avoid references to type tags in relocation
  entries.

[1] https://lore.kernel.org/bpf/87r0w9jjoq.fsf@oracle.com/

This was previously tracked as differential revision:
https://reviews.llvm.org/D145891

eddyz87 commented:
This is a part of the following commit stack:
- https://github.com/llvm/llvm-project/pull/91422
- https://github.com/llvm/llvm-project/pull/91423
- https://github.com/llvm/llvm-project/pull/91424

eddyz87 commented:
Hi @yonghong-song,

Could you please take a look at this pull request?
Rationale is here: https://github.com/llvm/llvm-project/pull/91422#issuecomment-2101104708

@dafaust, @jemarch, @4ast, fyi.

yonghong-song reviewed (approved):
LGTM. The change mostly the same as previous approved patch https://reviews.llvm.org/D145891 except adding some BTF Type Tag V1 handling.
I tried a few examples as well as bpf selftest with both V1 and V2. All tests are passed.

eddyz87 committed:
[BPF] '-S -emit-llvm -> -emit-llvm' for attr-btf_type_tag-*.c tests

Newly added btf_type_tag related tests use 'clang -S -emit-llvm',
which is now not accepted by 'clang'.
Fix these tests by using only '-emit-llvm'.

To be squashed with the main commit when when branch would be merged.
